### PR TITLE
Export test suite as single function

### LINKED ISSUE #157: Export simple interface to entire abstract suite
I'm working on a `localstorage-down` alternative (not because I need it, but was halfway done for PouchDB, and figured I see it through to get a new perspective on the whole buffer/browser/serialize story).

It cannot offer snapshot guarantees, which is fine if documented. But if you need to skip tests, like snapshots, you need to replace:

```js
require('abstract-leveldown/abstract/iterator-test').all(leveldown, test, testCommon)
```

With 5 lines more:

```js
require('abstract-leveldown/abstract/iterator-test').setUp(leveldown, test, testCommon);
require('abstract-leveldown/abstract/iterator-test').args(test);
require('abstract-leveldown/abstract/iterator-test').sequence(test);
require('abstract-leveldown/abstract/iterator-test').iterator(leveldown, test, testCommon, testCommon.collectEntries);
// require('abstract-leveldown/abstract/iterator-test').snapshot(leveldown, test, testCommon);
require('abstract-leveldown/abstract/iterator-test').tearDown(test, testCommon);
```

Or be clever about it:

```js
require('abstract-leveldown/abstract/iterator-test').snapshot = function() {}
require('abstract-leveldown/abstract/iterator-test').all(leveldown, test, testCommon)
```

Neither is ideal. Same story with other tests. In addition, some tests were missing from `localstorage-down`. No one would know.

So I was thinking: can we export a simple function from `abstract/index.js`, that runs all tests by default, takes the usual arguments, as well as options to exclude tests? Something like:

```js
require('abstract-leveldown/abstract')(leveldown, test, testCommon, {
  'open': {
    skip: 'openAdvanced'
  }
  'iterator': {
    skip: 'snapshot'
  },
  'approximate-size': {
    skip: 'all'
  }
})
```

Or move frequently skipped tests to separate files so we don't need nested options:

```js
require('abstract-leveldown/abstract')(leveldown, test, testCommon, {
  skip: [
    'open-advanced',
    'snapshot',
    'approximate-size'
  ]
})
```

Also makes it really easy to see what an implementation *doesn't* support.

@ralphtheninja @juliangruber

### PR DESCRIPTION
**Pull request by vweevers**
Do not merge, ~~I'm just exploring~~. This will make it much easier to update implementations, because they only have to call one function. @ralphtheninja WDYT?

Closes #157.

### CODE REVIEW

**Comment by ralphtheninja** [comment]
Hehe this goes in the direction of how `level-packager` used to be. But I actually like it because there are so many tests it's helpful.

**Comment by vweevers** [comment]
What do you like better, before or after 45dd8a8? On the one hand a single options object is cleaner but I don't like that it has required "options".

**Comment by ralphtheninja** [comment]
> What do you like better, before or after 45dd8a8? On the one hand a single options object is cleaner but I don't like that it has required "options".

After. Mostly because implementations don't need to explicitly require `test/common`, modify it and then pass it back in.

**Comment by vweevers** [comment]
> Mostly because implementations don't need to explicitly require test/common, modify it and then pass it back in.

Roger. In fact, we can remove `test/common` altogether..

**Comment by ralphtheninja** [comment]
> Roger. In fact, we can remove test/common altogether..

Aye, it could just be part of `test/index.js`.

**Comment by vweevers** [comment]
I want a gif for how everything is falling into place

**Comment by ralphtheninja** [comment]
falling-into-place

**Comment by vweevers** [comment]
... not what I was expecting 🤣

**Comment by ralphtheninja** [comment]
> Hmm, test.js does still need testCommon.

We could also consider duplicating the code, noop `setUp()` and `tearDown()` methods isn't a lot of code.

**Comment by vweevers** [comment]
I opted to move the options logic to `test/common` as a function, will push shortly.

**Comment by ralphtheninja** [comment]
@vweevers Will you close https://github.com/Level/abstract-leveldown/issues/264 in this PR as well?

**Comment by vweevers** [comment]
> Will you close #264 in this PR as well?

Sure!

File: test/index.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+var common = require('./common')
+
+/**
+ * Example usage:
+ *
+ * ```
+ * require('abstract-leveldown/test')({
```
Comment (vweevers) [inline test/index.js]: Except, this will resolve to `test.js`

Comment (vweevers) [reply test/index.js]: @ralphtheninja I just like `require('abstract-leveldown/test')` too much, so what's a good new name for `test.js`?

Comment (ralphtheninja) [reply test/index.js]: We can rename local `test.js` to something else.

Comment (vweevers) [reply test/index.js]: It should still match nyc/instanbul default ignore patterns. Maybe `test/self.js`?

Comment (ralphtheninja) [reply test/index.js]: Naming is always the hardest :smile: How about `test-local.js`?

Comment (vweevers) [reply test/index.js]: Should match both:

- `['**/node_modules/**', '**/bower_components/**', '**/test/**', '**/tests/**', '**/*.json']` (browserify-instanbul)
- `['coverage/**', 'test/**', 'test{,-*}.js', '**/*.test.js', '**/__tests__/**', '**/node_modules/**']` (nyc)

So the file must be in `test/`.

**Comment by vweevers** [comment]
I'll continue after a short break :) We're on a roll!

**Comment by vweevers** [comment]
> Maybe a short thing should go into README as well, a simple example of how to call the test function.

Can do that as part of #259

**Pull request merged**